server/src/infra/ldap_handler.rs

```diff
@@ -527,6 +544,47 @@ impl<Backend: BackendHandler + LoginHandler + OpaqueHandler> LdapHandler<Backend
         Ok(vec![make_add_error(LdapResultCode::Success, String::new())])
     }
 
+    async fn do_compare(&mut self, request: LdapCompareRequest) -> LdapResult<Vec<LdapOp>> {
+        let req = make_search_request::<String>(
+            &self.ldap_info.base_dn_str,
+            LdapFilter::And(vec![LdapFilter::Equality(
+                "dn".to_string(),
+                request.dn.to_string(),
+            )]),
+            vec![],
+        );
+        let res = self.do_search_or_dse(&req).await?;
+        let entry = res.first().ok_or(LdapError {
+            code: LdapResultCode::OperationsError,
+            message: "Search request returned nothing".to_string(),
+        })?;
+        match entry {
+            LdapOp::SearchResultEntry(entry) => {
+                let mut available = false;
+                for attr in entry.attributes.iter() {
+                    if attr.atype == request.atype && attr.vals.contains(&request.val) {
```

Comment:
I'm afraid that comparing the `atype` will not work if the comparison attribute is not normalized (i.e. requesting `distinguishedName` or even `DN` when we return `dn`).

It's also missing the case of "attribute is not part of the schema". I think we should start by checking whether the entry is a user or a group, then map both the compare attribute and the entry attributes using `map_user_field` (or `map_group_field`) before checking for equality. That'll also give us the case of "attribute not supported".

Reply:
You suggest also mapping the `atype`s on the entries, but aren't these already "normalized" as they're generated internally? I'm asking as this would save us some computation and make the code clearer. But I can totally do that if it's necessary.

Reply:
The `compare`'s `atype` is not normalized yet.

Reply:
What you can do is run the request's atype through a "LDAP attribute type -> SQL column -> LDAP attribute type" cycle to normalize it.